File: Frontend/src/app/core/services/payment-expense-link.service.ts
import { environment } from '../../../environments/environment';
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface PaymentExpenseLink {
  expenseId: number;
  eventGroupKey: number;
  eventGroupName?: string;
  vendorKey: number;
  vendorName?: string;
  expenseCategoryKey: number;
  categoryName?: string;
  expenseDetails?: string;
  billNo?: string;
  billDate?: string;
  amount: number;
  paidByMemberKey?: number;
  memberName?: string;
  paymentMode?: string;
  relatedPaymentID?: number;
  remarks?: string;
  isSettled: boolean;
  settledDate?: string;
  paymentId?: number;
  isVoucherOnly?: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class PaymentExpenseLinkService {
  private apiUrl = environment.apiUrl + '/PaymentExpenseLink';

  constructor(private http: HttpClient) { }

  getAll(): Observable<PaymentExpenseLink[]> {
    return this.http.get<PaymentExpenseLink[]>(this.apiUrl);
  }

  getPendingAll(): Observable<PaymentExpenseLink[]> {
    return this.http.get<PaymentExpenseLink[]>(`${this.apiUrl}/pending`);
  }

  getPendingByMember(memberId: number): Observable<PaymentExpenseLink[]> {
    return this.http.get<PaymentExpenseLink[]>(`${this.apiUrl}/pending/${memberId}`);
  }

  getUnlinkedExpenses(eventGroupKey: number, currentExpenseId?: number): Observable<PaymentExpenseLink[]> {
    let params = new HttpParams();
    if (currentExpenseId) {
      params = params.set('currentExpenseId', currentExpenseId.toString());
    }
    return this.http.get<PaymentExpenseLink[]>(`${this.apiUrl}/unlinked/${eventGroupKey}`, { params });
  }

  create(link: Partial<PaymentExpenseLink>): Observable<number> {
    return this.http.post<number>(this.apiUrl, link);
  }

  update(id: number, link: Partial<PaymentExpenseLink>): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}`, link);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: Backend/DemoProject.Business/Services/PaymentExpenseLinkService.cs
using DemoProject.Business.DTOs;
using DemoProject.Business.Interfaces;
using DemoProject.Data.Entities;
using DemoProject.Repository.Interfaces;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace DemoProject.Business.Services
{
    public class PaymentExpenseLinkService : IPaymentExpenseLinkService
    {
        private readonly IPaymentExpenseLinkRepository _repository;

        public PaymentExpenseLinkService(IPaymentExpenseLinkRepository repository)
        {
            _repository = repository;
        }

        public async Task<int> CreateAsync(CreatePaymentExpenseLinkDto dto, int createdBy)
        {
            var entity = new PaymentExpenseLink
            {
                EventGroupKey = dto.EventGroupKey,
                VendorKey = dto.VendorKey,
                ExpenseCategoryKey = dto.ExpenseCategoryKey,
                ExpenseDetails = dto.ExpenseDetails,
                BillNo = dto.BillNo,
                BillDate = dto.BillDate,
                Amount = dto.Amount,
                PaidByMemberKey = dto.PaidByMemberKey,
                PaymentMode = dto.PaymentMode,
                RelatedPaymentID = dto.RelatedPaymentID,
                Remarks = dto.Remarks
            };

            return await _repository.AddAsync(entity, createdBy);
        }

        public async Task UpdateAsync(UpdatePaymentExpenseLinkDto dto, int id, int modifiedBy)
        {
            var entity = new PaymentExpenseLink
            {
                ExpenseId = id,
                EventGroupKey = dto.EventGroupKey,
                VendorKey = dto.VendorKey,
                ExpenseCategoryKey = dto.ExpenseCategoryKey,
                ExpenseDetails = dto.ExpenseDetails,
                BillNo = dto.BillNo,
                BillDate = dto.BillDate,
                Amount = dto.Amount,
                PaidByMemberKey = dto.PaidByMemberKey,
                PaymentMode = dto.PaymentMode,
                RelatedPaymentID = dto.RelatedPaymentID,
                Remarks = dto.Remarks
            };

            await _repository.UpdateAsync(entity, modifiedBy);
        }

        public async Task DeleteAsync(int id, int modifiedBy)
        {
            await _repository.DeleteAsync(id, modifiedBy);
        }

        public async Task<IEnumerable<PaymentExpenseLinkDto>> GetAllAsync()
        {
            var entities = await _repository.GetAllAsync();
            return entities.Select(MapToDto);
        }

        public async Task<IEnumerable<PaymentExpenseLinkDto>> GetPendingAllAsync()
        {
            var entities = await _repository.GetPendingAllAsync();
            return entities.Select(MapToDto);
        }

        public async Task<IEnumerable<PaymentExpenseLinkDto>> GetPendingByMemberAsync(int memberId)
        {
            var entities = await _repository.GetPendingByMemberAsync(memberId);
            return entities.Select(MapToDto);
        }

        public async Task<IEnumerable<PaymentExpenseLinkDto>> GetUnlinkedExpensesAsync(int eventGroupKey, int? currentExpenseId)
        {
            var entities = await _repository.GetUnlinkedExpensesAsync(eventGroupKey, currentExpenseId);
            return entities.Select(MapToDto);
        }

        private PaymentExpenseLinkDto MapToDto(PaymentExpenseLink entity)
        {
            return new PaymentExpenseLinkDto
            {
                ExpenseId = entity.ExpenseId,
                EventGroupKey = entity.EventGroupKey,
                EventGroupName = entity.EventGroupName,
                VendorKey = entity.VendorKey,
                VendorName = entity.VendorName,
                ExpenseCategoryKey = entity.ExpenseCategoryKey,
                CategoryName = entity.CategoryName,
                ExpenseDetails = entity.ExpenseDetails,
                BillNo = entity.BillNo,
                BillDate = entity.BillDate,
                Amount = entity.Amount,
                PaidByMemberKey = entity.PaidByMemberKey,
                MemberName = entity.MemberName,
                PaymentMode = entity.PaymentMode,
                RelatedPaymentID = entity.RelatedPaymentID,
                Remarks = entity.Remarks,
                IsSettled = entity.IsSettled,
                SettledDate = entity.SettledDate,
                PaymentId = entity.PaymentId
            };
        }
    }
}
